src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -156,6 +215,663 @@ MultiRouterPlanCreate(Query *originalQuery, Query *query,
+	if (copiedInsertRte->alias == NULL)
+	{
+		Alias *alias = makeAlias(CITUS_TABLE_ALIAS, NIL);
+		copiedInsertRte->alias = alias;
+	}
+
+	/* and generate the full query string */
+	deparse_shard_query(copiedQuery, distributedTableId, shardInterval->shardId,
+						queryString);
+	ereport(DEBUG4, (errmsg("distributed statement: %s", queryString->data)));
+
+	modifyTask = CreateBasicTask(jobId, taskIdIndex, MODIFY_TASK, queryString->data);
+	modifyTask->dependedTaskList = NULL;
+	modifyTask->anchorShardId = shardId;
+	modifyTask->taskPlacementList = insertShardPlacementList;
+	modifyTask->upsertQuery = upsertQuery;
+
+	return modifyTask;
+}
+
+
+/*
+ * AddShardIntervalRangeToSelect adds the following range boundaries
+ * with the given subquery and shardInterval:
+ *
+ *    hashfunc(partitionColumn) >= $lower_bound AND
+ *    hashfunc(partitionColumn) >= $upper_bound
+ *
+ * The function expects and asserts that subquery's target list contains a partition
+ * column value.
+ */
+static void
+AddShardIntervalRangeToSelect(Query *subqery, ShardInterval *shardInterval)
+{
+	List *targetList = subqery->targetList;
+	ListCell *targetEntryCell = NULL;
+	Var *targetPartitionColumnVar = NULL;
+	Oid integer4GEoperatorId = InvalidOid;
+	Oid integer4LEoperatorId = InvalidOid;
+	TypeCacheEntry *typeEntry = NULL;
+	FuncExpr *hashFunctionExpr = NULL;
+	OpExpr *greaterThanAndEqualsBoundExpr = NULL;
+	OpExpr *lessThanAndEqualsBoundExpr = NULL;
+
+	/* iterate through the target entries */
+	foreach(targetEntryCell, targetList)
+	{
+		TargetEntry *targetEntry = lfirst(targetEntryCell);
+
+		if (IsPartitionColumnRecursive(targetEntry->expr, subqery) &&
+			IsA(targetEntry->expr, Var))
+		{
+			targetPartitionColumnVar = (Var *) targetEntry->expr;
+			break;
+		}
+	}
+
+	/* we should have found target partition column */
+	Assert(targetPartitionColumnVar != NULL);
+
+	integer4GEoperatorId = get_opfamily_member(INTEGER_BTREE_FAM_OID, INT4OID,
+											   INT4OID,
+											   BTGreaterEqualStrategyNumber);
+	integer4LEoperatorId = get_opfamily_member(INTEGER_BTREE_FAM_OID, INT4OID,
+											   INT4OID,
+											   BTLessEqualStrategyNumber);
+
+	/* look up the type cache */
+	typeEntry = lookup_type_cache(targetPartitionColumnVar->vartype,
+								  TYPECACHE_HASH_PROC_FINFO);
+
+	/* probable never possible given that the tables are already hash partitioned */
+	if (!OidIsValid(typeEntry->hash_proc_finfo.fn_oid))
+	{
+		ereport(ERROR, (errcode(ERRCODE_UNDEFINED_FUNCTION),
+						errmsg("could not identify a hash function for type %s",
+							   format_type_be(targetPartitionColumnVar->vartype))));
+	}
+
+	/* generate hashfunc(partCol) expression */
+	hashFunctionExpr = makeNode(FuncExpr);
+	hashFunctionExpr->funcid = typeEntry->hash_proc_finfo.fn_oid;
+	hashFunctionExpr->args = list_make1(targetPartitionColumnVar);
+
+	/* hash functions always return INT4 */
+	hashFunctionExpr->funcresulttype = INT4OID;
```

Comment:
Is that always correct? I think all our hashing infratructure based on this assumption, so I used that again.